OIDC tokens

SergeyYakubov opened the pull request:
#    Refresh and expose OIDC tokens
    
The idea is to be able to use Galaxy login via an OIDC provider as a single sign-on and
 authenticate when needed during workflow execution using OIDC tokens. "When needed"
may include storage access, compute nodes access, etc. Additionally, we make sure 
tokens are always fresh, so for each API request, check if tokens are not expired and refresh otherwise.
If the refresh fails, the user will be logged out. 


## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  1. Login using an OIDC provider (Keycloak, Azure AD, Google, etc.)
  2. Add expose a token to a tool, e.g.
  
        ```
           <environment_variables>
                <environment_variable name="_GALAXY_OIDC_ID_TOKEN" inject="oidc_id_token"/>
           </environment_variables>
        ``` 
  3. Check for `Refreshed user token...` message in log files (need to run Galaxy long enough for a token to expire).

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

SergeyYakubov committed:
Refresh and expose OIDC tokens

- Implement an option to expose OIDC tokens (if a user logged in
using OIDC protocol) to tools.

- At each API request check if
tokens are not expired and refresh otherwise.

- Logout user if refresh fails.

SergeyYakubov committed:
Update Python dependencies

SergeyYakubov committed:
remove unused imports

mvdbeek commented on lib/galaxy/webapps/base/webapp.py:
```diff
@@ -322,6 +323,13 @@ def __init__(
             self._ensure_valid_session(session_cookie)
+
+        if isinstance(self.user, User) and (self.user.social_auth or self.user.custos_auth):
+            success = self.app.authnz_manager.refresh(self)
```
I would assume that is too costly to do on every request. As a simple hack i'd maybe look at the galaxy_session update_time and refresh the token after n minutes (where you could make n configurable). When the token is refreshed you also bump the update_time.

Longer term I think it's probably a good idea to have a separate process that just refreshes tokens for logged-in users and logs out users if the refresh fails.

SergeyYakubov replied:
refresh would not happen at each request, inside the call we check if the token is about to expire and if not, we return.
https://github.com/galaxyproject/galaxy/pull/15300/files#diff-52d2ab08a1b269dab28ef1bb2feb00aef8eac9a200797994f34d1868b7de51c1R171
So, real refresh does happen every n minutes and expiration time is encoded in token, so no need to do it configurable.
To avoid confusion here, I could rename `refresh` -> `refresh_if_needed` or something like that

mvdbeek replied:
Nice, that's good to know, and yes, `refresh_if_needed` sounds good. I suppose you could still have many requests at once triggering the refresh, but I think solving this would require an app-wide user-specific lock mechanism (probably via the database), we can postpone that.

mvdbeek commented on lib/galaxy/tools/evaluation.py:
```diff
@@ -634,11 +634,28 @@ def _build_environment_variables(self):
                 else:
                     environment_variable_template = ""
                 is_template = False
+            elif inject == "oidc_refresh_token":
```
There could be an arbitrarily long wait time until the job starts and the values are actually used, is that going to be a problem ?

mvdbeek replied:
I think users can be logged in  / associated with multiple providers, do we need to anticipate the possibility that tools require "provider-specific" (google, azure, some other more specific service) tokens ?

SergeyYakubov replied:
> There could be an arbitrarily long wait time until the job starts and the values are actually used, is that going to be a problem ?

It depends, this is actually an issue for `access_token` and `id_token`, this is why we also expose `refresh_token`, which usually has a much longer expiration time (24+ hours), and it depends on the identity provider. If a job starts after that, there is hardly much we can do - single sign-on would not work. I was thinking about adding a Galaxy API to get a fresh token, but we have to see if there are security considerations and really see the use case where a job start takes that long.

SergeyYakubov replied:
> I think users can be logged in / associated with multiple providers, do we need to anticipate the possibility that tools require "provider-specific" (google, azure, some other more specific service) tokens ?

This is a good catch, thank you! I'm actually always using the first token,
https://github.com/galaxyproject/galaxy/pull/15300/files#diff-59a534d07c0974fda8548d41482653dd5512d69681950449207ec5c0f1ccc3f3R757
but if there are more than one, that might not work. We'd need to get the one from the provider that was used to login.

SergeyYakubov replied:
@mvdbeek - I do not see a nice way to get information about the provider used for login. I think it is not stored in Galaxy, but please correct me if I'm wrong.

hexylena replied:
xref https://github.com/galaxyproject/galaxy/issues/14578, yes, we absolutely need to specify the provider. There I proposed a tool syntax like:

```
    <environment_variables>
        <environment_variable name="LS_TOKEN" provider="elixir" inject="access_token" />
    </environment_variables>
```

since in many cases the tool is specific to one endpoint and will only accept one flavour of token.

SergeyYakubov replied:
Hm, we've been thinking about the situation where Galaxy is configured to work with multiple providers. For us, it is kind of an edge case when the same user logs in using multiple providers. Why would that happen? And what is the expected behavior? Should we automatically update all tokens? What should we do if one of the updates fails, but another is ok - logout user or not?
So, what I currently implemented - if a user logs in using one provider, tokens from all other providers are invalidated. So there can only be one set of active tokens from a specific provider. Example - a user logs in using Elixir, then the same user logs in using Google - Elixir tokens will be deleted, only Google ones will be refreshed.

As for your @hexylena idea to specify the provider in a tool - this is a good one! But I'd suggest making it optional, because some tools might just need a token (to pass it further to some service) and mustn't know what kind of token it is. So, if provider is specified - we check if such token is active and return it, if no provider specified - we just return an active token.


Does this all make sense?

hexylena replied:
> For us, it is kind of an edge case when the same user logs in using multiple providers.

I somewhat agree here, it's probably relatively rare.

> Why would that happen? And what is the expected behavior? Should we automatically update all tokens? What should we do if one of the updates fails, but another is ok - logout user or not?

I guess the question is: how common is the case where a tool needs a specific token? (I guess this will support two cases: tools and built-in functionality, but focusing on tools for a moment.)

For the tools I have in mind they all need a specific provider, and if another provider token update fails that's absolutely fine. I don't think token refreshing should log out the user immediately, I don't know if that's a valid thing to do (does a failed refresh strictly imply loss of access to the account? e.g. in a temporary backend failure it wouldn't make sense to kick the user.)

For the built-in functionality case like accessing remote data / storing datasets to a user provided backend, then, yeah, maybe makes a lot more sense to kick them immediately since everything will break.

> So, what I currently implemented - if a user logs in using one provider, tokens from all other providers are invalidated. So there can only be one set of active tokens from a specific provider. Example - a user logs in using Elixir, then the same user logs in using Google - Elixir tokens will be deleted, only Google ones will be refreshed.

Interesting! I can see why, but, yeah for the specific elixir case, there's a tool the user wants to execute which requires elixir authentication. It would be absolutely fine then to e.g. store that data on google or mix that data with data from google, and having them go through a logout/login flow just to execute a tool seems rather unpleasant for the user?

> But I'd suggest making it optional, because some tools might just need a token (to pass it further to some service) and mustn't know what kind of token it is.

yes, making it optional is fine for me.

(but personally I'm curious what kind of tools can use a token completely interchangeably, wouldn't this also require selecting different API endpoints or so?)

SergeyYakubov replied:
> , I don't know if that's a valid thing to do (does a failed refresh strictly imply loss of access to the account? e.g. in a temporary backend failure it wouldn't make sense to kick the user.)

I think we should kick the user. If one logged in through some provider and we cannot refresh (read re-verify) the user, we cannot trust this user anymore. So, I think we should log them out. One thing we can do is to add a grace period -  since we refresh tokens before the access token expires anyway, we can return ok in case of failure for some time and only return error after the original access_token expires.


> It would be absolutely fine then to e.g. store that data on google or mix that data with data from google, and having them go through a logout/login flow just to execute a tool seems rather unpleasant for the user?

I'm not sure I quite understand that. So a user logs in via Google, does something with data, then logs out, and logs in with Elixir. But, in spite of logout, do we still want to keep Google tokens and refresh them? And if that Google account does not exist, are we still ok?

I can try to implement that (as I said, for us that should never happen, so we should be good) if everybody agrees. In this case, we'd be refreshing all tokens, never logout a user if refresh fails, and asking for a provider in a tool. 


> (but personally I'm curious what kind of tools can use a token completely interchangeably, wouldn't this also require selecting different API endpoints or so?)

we have such a tool - it grabs some data via scp, using OIDC token to login - so no need to know what this token is, we just pass it to sshd service on the other side.


Btw, custos auth does not allow multiple providers (so you cannot have Keycloak and Google together for the same user).

hexylena replied:
> I think we should kick the user. If one logged in through some provider and we cannot refresh (read re-verify) the user, we cannot trust this user anymore. So, I think we should log them out. One thing we can do is to add a grace period - since we refresh tokens before the access token expires anyway, we can return ok in case of failure for some time and only return error after the original access_token expires.

Yeah ok. I can agree to it. I'm worried about edge cases, but, maybe they're just that: edge cases.

> I'm not sure I quite understand that. So a user logs in via Google, does something with data, then logs out, and logs in with Elixir. But, in spite of logout, do we still want to keep Google tokens and refresh them? And if that Google account does not exist, are we still ok?

Yeah ok, maybe again I'm too worried about multi-auth'd accounts. I know we'll have users that login via elixir and never use their token, and other users that login explicitly to use that token in the tool which consumes it.

I guess I'm working under the impression that by also having an e.g. Google account linked, in addition to their elixir account, that they can do different things with that. That they'll never explicitly click "logout", but instead:

- login with elixir
- attach google as well via user preferences (rather than logging out and logging in with google which would generate a second galaxy account because they don't resolve the same way)
- do neat things that require tokens from both

> In this case, we'd be refreshing all tokens, never logout a user if refresh fails, and asking for a provider in a tool.

yeah, in **this** case _not_ kicking the user makes more sense.
But if they only logged in via google, and the token fails to validate, then I agree we should kick them.

For the _vast majority_ of our users, they had an existing account, and they attached a social login to that to get some benefit, rather than primarily logging in with social auth. 

But for your case where they've only ever logged in via social, hard agree, kick if validation fails. But those users won't have proper passwords right? So many the logic comes down to, in the event of validation failure

- if only one social
  - and no password ⇒ kick
  - and a password ⇒ they can still login, don't kick
- if multiple socials
  - and they all fail to validate ⇒ kick
  - and they have a password ⇒ don't kick.

> we have such a tool - it grabs some data via scp, using OIDC token to login - so no need to know what this token is, we just pass it to sshd service on the other side.

huh, neat!

> Btw, custos auth does not allow multiple providers (so you cannot have Keycloak and Google together for the same user).

good reason for me to not switch to custos :)

SergeyYakubov replied:
ok, I think I start to understand :). Thank you for the clarifications!

So, we basically have two scenarios here. The first one is when a user logs in using social auth. In this case one would not normally expect tokens from multiple providers and would expect token verification at each request and logout if something is wrong. This is the scenario I had in mind.

A second scenario is when an existing user connects to a third-party IDP using user preferences. I was not even aware of that option before :(, but I guess this is more your use case. In this case you indeed do not want to logout user and you do want to select which token to expose in tools.

Ideally, we would distinguish these two scenarios, keeping track of which provider a user used to log in and which providers they connected extra. Then we could return the "provider-last-used-to-login" token as default, kick off a user only if the "provider-last-used-to-login" token cannot be refreshed (if there is such a token), and refresh all other tokens.
Should I go forward and implement that? It would require adding a field to the user table in the database, which I tried to avoid before.

Alternatively, as I said before, we can just refresh everything and never kick users. Worst case - we allow a user to be logged in with an expired token. It would work for us since they would not be able to access data/resources.

dannon replied:
> A second scenario is when an existing user connects to a third-party IDP using user preferences. I was not even aware of that option before :(, but I guess this is more your use case. In this case you indeed do not want to logout user and you do want to select which token to expose in tools.

> Ideally, we would distinguish these two scenarios, keeping track of which provider a user used to log in and which providers they connected extra. Then we could return the "provider-last-used-to-login" token as default, kick off a user only if the "provider-last-used-to-login" token cannot be refreshed (if there is such a token), and refresh all other tokens. Should I go forward and implement that? It would require adding a field to the user table in the database, which I tried to avoid before.

Would an explicit credential selector input that gets rendered to the tool form resolve this?  In the first situation, with only one credential, it wouldn't allow you to change anything but it would make it clear to the user that a tool would be using their creds -- showing this explicitly seems like it'd be a good thing to do regardless.  In the second situation, you could provide a selector for the specific credential set to use, toggling between google and azure, or whatever you may have.

hexylena replied:
> Alternatively, as I said before, we can just refresh everything and never kick users. Worst case - we allow a user to be logged in with an expired token. It would work for us since they would not be able to access data/resources.

I think given the variety of galaxy deployments, and the fail-closed situation with your deployment, this is probably the "safest" and most user friendly option.

> Would an explicit credential selector input that gets rendered to the tool form resolve this?

oh that's a great idea, like a resource selector. Love this!

SergeyYakubov replied:
> Would an explicit credential selector input that gets rendered to the tool form resolve this?

Can we dynamically build an input list for this selector? The values should depend on the currently attached identities for a specific user. I'm not sure how to implement this. Could you please give me a hint?

mvdbeek replied:
We can probably attach the available credential source to the user object, and use that when we see "inject credentials" in a tool. Since it seems like you've got your usecase working I would suggest this for a followup PR.

mvdbeek reviewed (commented):
This looks very interesting, thanks so much @SergeyYakubov! I left some comments inline, but I don't know much about OIDC, I hope @hexylena and @nuwang have some input here.

SergeyYakubov committed:
fix lint error in lib/galaxy/tools/evaluation.py

nuwang commented on lib/galaxy/tools/evaluation.py:
```diff
@@ -639,6 +639,24 @@ def _build_environment_variables(self):
                 is_template = False
+            elif inject == "oidc_refresh_token":
+                if self._user:
+                    environment_variable_template = self._user.oidc_refresh_token
```
Will tools be using the refresh token? My understanding is that for a refresh token to be used to generate a new access token, the client id and secret must also be available.

SergeyYakubov replied:
> 1. Should we consider adding an allowlist to gate token injection, similar to how sanitize_allowlist is used on the client side?

What is the scenario for a security breach? Someone installs a malicious tool in Galaxy? Or does a normal tool runs in an untrusted environment? Or both? In the first case, sanitize_allowlist might help. In the second  - it would probably be better to have an allowed list of destinations for a tool. But this would complicate administration, I guess. 

I saw that there are no restrictions to expose Galaxy API tokens, so I just added OIDC tokens in the same way. Actually, losing the  API token is worse than an OIDC one since that one is time unlimited. 

> 2. What are the implications of token expiry during tool execution, something that's more likely to happen in high-load sites where jobs may sit for a long time in the queue.

Agree, this is a weak point in the whole scenario. A tool that needs to use OIDC tokens should not sit in a queue for too long. This is also why we expose a refresh token. But then, as you've noticed, you need to know the client id and secrets. We use it for one tool, passing corresponding values from the Galaxy config file (https://github.com/PaNGalaxy/neutron-tools/blob/main/tools/neutrons/import_oncat.xml#L18). Btw, the fact that we can do that is another security issue where one can write a tool that grabs Galaxy config from a server!

> 
> I would like to see this feature being available especially in trusted environments, where the entire stack, inclusive of the job execution environment, tools themselves, and everything in between, is trusted and secure. However, I think it's a feature that may need some access controls in other environments where there's less confidence in the tools in particular.

Agree, we developed that with our environment in mind, where everything is trusted and is under our control. So, would adding sanitize_allowlist be good enough as a first step? At least with that, by default, the feature is not available, and then one can switch it on for specific tools.

nuwang replied:
After thinking about this a bit more, I don't think my original concerns are relevant. While it would be nice to be able to enforce controls on specific tools, I think that a general permissions model for tools would be a far different undertaking and is not really within the scope of this PR. As you've pointed out, there are too many ways to bypass those things anyway, and we would probably need something like an oauth2 consent page at the point of installing the tool for a tool to request the specific permissions it needs and for admins to approve it. Therefore, the approach you've taken seems ok to me.

nuwang commented on lib/galaxy/authnz/custos_authnz.py:
```diff
@@ -67,6 +67,44 @@ def __init__(self, provider, oidc_config, oidc_backend_config, idphint=None):
+        )
+        if custos_authnz_token is None:
+            raise exceptions.AuthenticationFailed("cannot find authorized user while refreshing token")
+        if (custos_authnz_token.expiration_time - datetime.now()).total_seconds() > 60:
```
The refresh condition for this appears to be different for the PSA case: https://github.com/galaxyproject/galaxy/pull/15300/files#diff-52d2ab08a1b269dab28ef1bb2feb00aef8eac9a200797994f34d1868b7de51c1R171

Should they be the same?

SergeyYakubov replied:
The reason is that `CustosAuthnzToken` and `UserAuthnzToken` have different fields, in particular `CustosAuthnzToken` does not contain info about auth time, so I could not refresh custos tokens at their half-life (at least not without changing the database scheme).

nuwang reviewed (commented):
Thanks for taking this up @SergeyYakubov, this is very useful. While I would like to see support for the end-scenario you're working towards (allowing tools to perform oidc access on behalf of the user), I guess the two main concerns I had were:

1. Should we consider adding an allowlist to gate token injection, similar to how sanitize_allowlist is used on the client side?
2. What are the implications of token expiry during tool execution, something that's more likely to happen in high-load sites where jobs may sit for a long time in the queue.

I would like to see this feature being available especially in trusted environments, where the entire stack, inclusive of the job execution environment, tools themselves, and everything in between, is trusted and secure. However, I think it's a feature that may need some access controls in other environments where there's less confidence in the tools in particular.

SergeyYakubov committed:
fix oidc token refresh and expose in case of multiple providers

when multiple identity providers are configures and a user logs using one of them,
the tokens for the other providers are invalidated (deleted) so that we always expose
tokens from the latest used provider.

SergeyYakubov committed:
refresh and expose all oidc tokens

refresh oidc tokens from all connected providers,
do not logout user in case of refresh failure
use provider name in token inject keywords
refactor relevant code

SergeyYakubov committed:
Merge branch 'dev' into oidc_tokens

SergeyYakubov committed:
Merge branch 'dev' into oidc_tokens

SergeyYakubov committed:
fix lint and failing tests

SergeyYakubov committed:
fix mypy and unit tests for app.authnz_manager

mvdbeek commented on lib/galaxy/tools/evaluation.py:
```diff
@@ -639,6 +640,9 @@ def _build_environment_variables(self):
                 else:
                     environment_variable_template = ""
                 is_template = False
+            elif inject.startswith("oidc_"):
```
```suggestion
            elif inject and inject.startswith("oidc_"):
```
since `inject` can be `None`

SergeyYakubov committed:
fix inject=None case while building tool environment variables

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

mvdbeek commented on lib/galaxy/tool_util/parser/xml.py:
```diff
@@ -10,6 +10,7 @@
 
 import packaging.version
 
+from galaxy.authnz.util import provider_name_to_backend
```
Can you think of a way to avoid this import here and ideally also in evaluation.py ?
We want to be able to reason about tools in these modules without considering the server configuration, and this is breaking the package tests.

I guess one way would be to do the `provider_name_to_backend` check when building the concret tool class in https://github.com/mvdbeek/galaxy/blob/7eb3260e589c28cc046a606375355c76aa0df653/lib/galaxy/tools/__init__.py#L563

SergeyYakubov replied:
I see what you mean - tool validation should not depend on server config. I removed that part, only checking now that the syntax is correct:  `oidc_[id|access|refresh]_token_xxx`. If we cannot expose a token, we just set the corresponding environment variable to `token-unavailable`. This would happen now if a) `xxx` is not configured in Galaxy (or misspelled in a tool file) or b) a user has not logged in via this provider. 

For evaluation.py - I'm not sure how can we avoid calling `provider_name_to_backend ` to map the provider name to the backend name (e.g. `azure`->`azuread-v2-tenant-oauth2`). We could request to use the backend name in a tool XML file, but this is probably not a good idea. Or we could change the scheme and store the provider name in 
the database entry for OIDC tokens. Not sure if that is a good idea as well. But without any of these two options we need this mapping.

mvdbeek replied:
The code in evaluation.py may run on the compute node not having a database available, but we do build the and serialize the environment variables while we still have a database connection, so that's actually fine for now. We may later move this into the JobWrapper class, but it's fine for now.

SergeyYakubov committed:
do not validate OIDC provider while parsing a tool file

SergeyYakubov committed:
return OIDC tokens as dict

SergeyYakubov committed:
Merge branch 'dev' into oidc_tokens

SergeyYakubov committed:
add msal package dependency (for auth via azure ad)

SergeyYakubov committed:
fix linting

SergeyYakubov committed:
Merge branch 'dev' into oidc_tokens

SergeyYakubov committed:
Add an API endpoint to access OIDC tokens from Pulsar as discussed in #15526

SergeyYakubov committed:
Merge branch 'dev' into oidc_tokens

SergeyYakubov committed:
fix lint

mvdbeek commented on lib/galaxy/webapps/base/webapp.py:
```diff
@@ -322,6 +322,12 @@ def __init__(
+        try:
+            self.app.authnz_manager.refresh_expiring_oidc_tokens(self)  # type: ignore[attr-defined]
+        except Exception:
+            pass
```
Given the broad exception catching in `refresh_expiring_oidc_tokens` I don't think you can run into an exception here ?

SergeyYakubov replied:
The problem was that `app.authnz_manager` does not always exist (e.g. in tests), so I had to deal with that. I can, of course, just check that `app` has `authnz_manager ` member it that makes more sense.

mvdbeek commented on lib/galaxy/webapps/galaxy/api/job_tokens.py:
```diff
@@ -0,0 +1,78 @@
+log = logging.getLogger(__name__)
+
+
+class JobTokensAPIController(BaseGalaxyAPIController):
```
Can you write this as a FastAPI route ? We don't want to add new old-style API controllers anymore.
Here's an example of a relatively simple route: https://github.com/mvdbeek/galaxy/blob/41d93e9c80b00fda7bb464055228800bc8be7f64/lib/galaxy/webapps/galaxy/api/roles.py#L39

mvdbeek commented on lib/galaxy/webapps/galaxy/buildapp.py:
```diff
@@ -1075,6 +1075,14 @@ def connect_invocation_endpoint(endpoint_name, endpoint_suffix, action, conditio
+        action="get_token",
+        conditions=dict(method=["GET"]),
+    )
+
```
You can remove this when you port the API to FastAPI

mvdbeek commented on lib/galaxy/tools/evaluation.py:
```diff
@@ -98,6 +100,7 @@ def global_tool_logs(func, config_file, action_str):
 class ToolEvaluator:
     """An abstraction linking together a tool and a job runtime to evaluate
     tool inputs in an isolated, testable manner.
+    tool inputs in an isolated, testable manner.
```
```suggestion
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/job_tokens.py:
```diff
@@ -0,0 +1,78 @@
+        try:
+            for plugin in self.app.job_config.runner_plugins:
+                if plugin["id"] == job.job_runner_name:
+                    secret = plugin["kwds"]["secret"]
```
This should probably operate on the destination instead of the job runner plugin level, since this doesn't really seem tied to the job runner (even if it only works for pulsar) ~~and more tied to e.g. the user~~.
Also could `secret` be more descriptive ? maybe `job_oidc_token_secret_base` ?

SergeyYakubov replied:
Makes sense, thank you. Just one comment -  this secret can be used in other places for calling Galaxy API from some service. I.e. to get. job files. Currently that secret is hard-coded in the [source](https://github.com/galaxyproject/galaxy/blob/2be2cec8c5bee607cf04fc279f173af82d5f3627/lib/galaxy/webapps/galaxy/api/job_files.py#L124), which is not so nice. So I suggest to use `job_secret_base` instead of `job_oidc_token_secret_base `

mvdbeek commented on lib/galaxy/webapps/galaxy/api/job_tokens.py:
```diff
@@ -0,0 +1,78 @@
+    @expose_api_raw_anonymous_and_sessionless
+    def get_token(self, trans, job_id, **kwargs):
+        """
+        GET /api/jobs/{job_id}/oidc-tokens
```
Since you're just getting a single token `/api/jobs/{job_id}/oidc-token` might be preferable ?

SergeyYakubov replied:
My thoughts were that we have multiple tokens and get only one due to the query parameter (`provider`).
So we basically call  `/api/jobs/{job_id}/oidc-tokens` with a filter (we do not return all tokens if no filter is provided, though).

But I can change it to `oidc-token` if this is a general convention in Galaxy, no problem.

mvdbeek commented on lib/galaxy/webapps/galaxy/api/job_tokens.py:
```diff
@@ -0,0 +1,78 @@
+        :type   job_id: str
+        :param  job_id: encoded id string of the job
+        :type   provider: str
+        :param  provider: Path to file.
```
That seems like a copy-paste mistake ?

mvdbeek commented on lib/galaxy/model/__init__.py:
```diff
@@ -923,6 +923,40 @@ def adjust_total_disk_usage(self, amount, quota_source_label):
+                return auth
+        return None
+
+    def get_oidc_tokens(self, provider_backend):
```
Should this be `get_oidc_token` since you're just getting a single token ?

SergeyYakubov replied:
I return a dictionary with three tokens - _access_, _refresh_ and _id_

nsoranzo commented on lib/galaxy/webapps/base/webapp.py:
```diff
@@ -322,6 +322,12 @@ def __init__(
+        try:
+            self.app.authnz_manager.refresh_expiring_oidc_tokens(self)  # type: ignore[attr-defined]
+        except Exception:
+            pass
```
Yes, I think that would be more explicit and not hide other real exceptions.

```suggestion
        if hasattr(trans.app, "authnz_manager"):
            self.app.authnz_manager.refresh_expiring_oidc_tokens(self)
```

SergeyYakubov committed:
switch to FastAPI for job_tokens

SergeyYakubov committed:
use job_secret_base from destination parameters to secure an endpoint for OIDC tokens

SergeyYakubov committed:
remove superfluous exception

SergeyYakubov committed:
cleanup docstring

SergeyYakubov committed:
Merge branch 'dev' into oidc_tokens

SergeyYakubov committed:
restore missing dependency

SergeyYakubov committed:
update API scheme

SergeyYakubov committed:
fix check for app.authnz_manager
